src/Compilers/CSharp/Test/Semantic/Semantics/UsingDeclarationTests.cs

```diff
@@ -0,0 +1,396 @@
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label1;").WithLocation(7, 9),
+                // (8,9): warning CS0162: Unreachable code detected
+                //         using var x = (IDisposable)null;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "using").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void DisallowMultipleGoToForwardAcrossMultipleUsingDeclarations()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        goto label1;
+        using var x = (IDisposable)null;
+        goto label2;
+        using var y = (IDisposable)null;
+
+        label1:
+        label2:
+        return;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (7,9): error CS8641: A goto target within the same block can not cross a using declaration.
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label1;").WithLocation(7, 9),
+                // (8,9): warning CS0162: Unreachable code detected
+                //         using var x = (IDisposable)null;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "using").WithLocation(8, 9),
+                // (9,9): error CS8641: A goto target can not be after any using declarations.
+                //         goto label2;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label2;").WithLocation(9, 9)
+                );
+        }
+
+        [Fact]
+        public void DisallowGoToBackwardsAcrossUsingDeclarations()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        label1:
+        using var x = (IDisposable)null;
+
+        goto label1;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (10,9): error CS8641: A goto target within the same block can not cross a using declaration.
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToBackwardJumpOverUsingVar, "goto label1;").WithLocation(10, 9)
+                );
+        }
+
+        [Fact]
+        public void AllowGoToAroundUsingDeclarations()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        goto label1;
+        label1:
+        using var x = (IDisposable)null;
+        goto label2;
+        label2:
+        return;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AllowGoToAcrossUsingDeclarationsBackwardsWhenFromADifferentBlock()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        label1:
+        using var x = (IDisposable)null;
+        {
+            if (true)
+            {
+                goto label1;
```

Comment:
>goto label1; [](start = 16, length = 12)

Why is this allowed? If the `using` is crossed once, are two instances disposed or one? Please add a test that executes `Dispose` in such a scenario.

Reply:
Yes. We cross twice and dispose twice. It was explicitly allowed by LDM. Will add a test to validate multi-disposal.

Reply:
This is actually already covered by [CodeGenUsingDeclarationTests.JumpBackOverUsingDeclaration ](https://github.com/dotnet/roslyn/pull/31674/files#diff-089d974ab4dbb1a5396677a34836e075R980)

Reply:
From the [LDM notes](https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-12-03.md), it seems the above example should be an error:
> When `goto`ing a label in the same statement list as a `using var`, the `goto` must not have a `using var` syntactically between the `goto` and the label.

Reply:
We should check with LDM, but my understanding from the meeting was that 'statement list' meant it was on the same level in the block. We discussed that jumping back from a lower block should be allowed.

Reply:
Discussed offline and I was wrong here. I've updated the algorithm to handle it correctly: backwards jumps are now disallowed when the label and the using declaration are part of the same block.